File: 206reverseLinkedList.js
/**
 * @param {ListNode} head
 * @return {ListNode}
 */
var reverseList = (head) => {
    if(head == null || head.next == null)
        return head

    // sep list into head and remaining
    const rest = reverseList(head.next)
    head.next.next = head
    head.next = null
    return rest
};


let four = {
    val: 4,
    next: null
}
  
let three = {
    val: 3,
    next: four
}
  
let two = {
    val: 2,
    next: three
}

let one = {
    val: 1,
    next: two
}

console.log(reverseList(one))

File: 206reverseLinkedList.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def reverseList(self, head: ListNode) -> ListNode:
        curr = head
        prev = None
        
        while curr:
            # need to keep track of next to set to curr 
            # since we lose this
            next = curr.next 
            
            # swap next with prev
            curr.next = prev
            prev = curr
            curr = next
        
        # we return prev, since curr is null end of list
        return prev
                
